### PR TITLE
emacs: fix emacs packages with new fetchers

### PR DESCRIPTION
**Pull request by osama-re**
###### Description of changes

#190863 fixed the updating problem but broke those packages.

###### Things done

- handle melpa new fetchers in libgenerated.nix
- updated melpa packages
- updated package-build (needed for building the new fetchers recipes)
- patched package-build to not use mtime (set to 0) when creating tarballs before building the package (needed after melpa/package-build@f032c80)

- tested
  - sourcehut fetcher (new)
    - autocrypt (multi-file package)
    - better-defaults (single-file package)
    - bbdb-csv-import (single-file package)
    - buffer-sets (single-file package)
  - codeberg fetcher (new)
    - chronometrist (multi-file package)
    - doc-show-inline (single-file package)
  - github
    - magit

- Built on platform(s)
  - [X] x86_64-linux

### CODE REVIEW

File: pkgs/build-support/emacs/melpa.nix [hunk 1]
```diff
@@ -35,11 +35,24 @@ import ./generic.nix { inherit lib stdenv emacs texinfo writeText gcc; } ({
...
+      owner = "melpa";
+      repo = "package-build";
+      rev = "c3c535e93d9dc92acd21ebc4b15016b5c3b90e7d";
+      sha256 = "17z0wbqdd6fspbj43yq8biff6wfggk74xgnaf1xx6ynsp1i74is5";
+    };
+
+    patches = [ ./package-build-dont-use-mtime.patch ];
+
+    dontConfigure = true;
+    dontBuild = true;
```
Comment (yuuyins) [inline pkgs/build-support/emacs/melpa.nix:50]: if you remove that, does it success? afaik generally don't do that if not really needed.

```diff
-    dontConfigure = true;
-    dontBuild = true;
```

Comment (osama-re) [reply pkgs/build-support/emacs/melpa.nix:50]: No, it doesn't succeed. it needs emacs as a build input, and i don't know if that's okay.
The previous approach was to just copy the repo and load it to emacs, so i didn't want to change that.

File: pkgs/applications/editors/emacs/elisp-packages/libgenerated.nix [hunk 2]
```diff
@@ -45,6 +45,18 @@ let
           inherit sha256 url;
         }
       ) {};
+      sourcehut = self.callPackage ({ fetchzip }:
```
Comment (AndersonTorres) [inline pkgs/applications/editors/emacs/elisp-packages/libgenerated.nix:48]: Was this script-generated?

Comment (AndersonTorres) [reply pkgs/applications/editors/emacs/elisp-packages/libgenerated.nix:48]: Also, is it compatible with the generator scripts `update-*`?

Comment (osama-re) [reply pkgs/applications/editors/emacs/elisp-packages/libgenerated.nix:48]: > Was this script-generated?
No, i added it manually.

> Also, is it compatible with the generator scripts update-*?
compatible how? the update-* scripts don't touch libgenerated.nix?

**Comment by ajgrf** [comment]
Can the new melpa fetchers be backported to nixos-22.05? I'm pretty sure it would unbreak emacs-overlay for me. This is what I'm getting when I try to update emacs with my packages:

```
error: attribute 'sourcehut' missing for call to 'getAttr'

       at /nix/store/4381sdcm40wvdh7hslc8jjrnjn4b022l-source/pkgs/applications/editors/emacs/elisp-packages/libgenerated.nix:10:33:

            9|                         , commit
           10|                         , ...}: {
             |                                 ^
           11|       github = self.callPackage ({ fetchFromGitHub }:
(use '--show-trace' to show detailed location information)
```

**Comment by AndersonTorres** [comment]
Remember me three days and I will try this

**Comment by ajgrf** [comment]
Just a reminder, it's been 3 days. :smile:

**Comment by AndersonTorres** [comment]
@ajgrf 

https://github.com/NixOS/nixpkgs/pull/193992

**Pull request merged**